src/wallet/wallet.cpp

```diff
@@ -857,80 +858,128 @@ bool CWallet::LoadUnifiedAddressMetadata(const ZcashdUnifiedAddressMetadata &add
-        return std::visit(match {
-            [&](const CKeyID& addr) {
-                auto ua = self->FindUnifiedAddressByReceiver(addr);
-                if (ua.has_value()) {
-                    return libzcash::PaymentAddress{ua.value()};
-                } else {
-                    return libzcash::PaymentAddress{addr};
-                }
-            },
-            [&](const CScriptID& addr) {
-                auto ua = self->FindUnifiedAddressByReceiver(addr);
-                if (ua.has_value()) {
-                    return libzcash::PaymentAddress{ua.value()};
-                } else {
-                    return libzcash::PaymentAddress{addr};
-                }
-            },
-            [&](const SaplingPaymentAddress& addr) {
-                auto ua = self->FindUnifiedAddressByReceiver(addr);
-                if (ua.has_value()) {
-                    return libzcash::PaymentAddress{ua.value()};
-                } else if (ufvk.has_value() && ufvk->GetSaplingKey().has_value()) {
-                    auto saplingKey = ufvk->GetSaplingKey().value();
-                    auto j = saplingKey.DecryptDiversifier(addr.d);
-                    // std::get is safe here because we know we have a valid Sapling diversifier index
-                    auto defaultUA = std::get<std::pair<UnifiedAddress, diversifier_index_t>>(
-                            ufvk->Address(j, CWallet::DefaultReceiverTypes()));
-                    return libzcash::PaymentAddress{defaultUA.first};
-                } else {
-                    return libzcash::PaymentAddress{addr};
+
+    int nHeight = chainActive.Height();
+    auto wtxPtr = mapWallet.find(txid);
+    if (wtxPtr != mapWallet.end()) {
+        const CBlockIndex* pTxIndex{nullptr};
+        if (wtxPtr->second.GetDepthInMainChain(pTxIndex) > 0) {
+            nHeight = pTxIndex->nHeight;
+        }
+    }
+
+    auto ufvk = self->GetUFVKForReceiver(RecipientAddressToReceiver(recipient));
+    std::pair<PaymentAddress, RecipientType> defaultAddress = std::visit(match {
+        [&](const CKeyID& addr) {
+            auto ua = self->FindUnifiedAddressByReceiver(addr);
+            if (ua.has_value()) {
+                // we do not generate unified addresses for change keys
+                return std::make_pair(PaymentAddress{ua.value()}, RecipientType::WalletExternalAddress);
+            }
+
+            // not associated with an UA, so we check legacy key metadata; if we don't
+            // have any metadata, it's a counterparty address
+            auto keyMeta = mapKeyMetadata.find(addr);
+            if (keyMeta == mapKeyMetadata.end()) {
+                return std::make_pair(PaymentAddress{addr}, RecipientType::CounterpartyAddress);
+            }
+
+            // for legacy change keys, we can check the keypath to determine
+            // whether this is an internal or external address.
+            if (IsInternalKeyPath(44, BIP44CoinType(), keyMeta->second.hdKeypath)) {
```

Comment:
`hdKeypath` will be empty if the address was generated before 4.7.0 (i.e. randomly, not from a seed via BIP 44).

Reply:
Fixed in https://github.com/zcash/zcash/pull/5715/commits/151074cbb13455afb8257354c0d856a79da3f332